regression: ast gen fails to catch incorrect by ref error captures

### Zig Version

0.12.0-dev.2158+4f2009de1

### Steps to Reproduce and Observed Behavior

this is a follow up to #18579 and addresses some changes made in #18173.  

these code snippets below were produced by @Techatrix and i think he may address them soon.  we noticed this while working on #18579 but decided that it deserves its own issue.

with 0.11.0 the following produced a compile error `error set cannot be captured by reference`.  but now the compiler allows it.  i think this is a regression.  
```zig
comptime {
    const e: error{A}!u32 = error.A;

    e catch |err| switch (err) {
        error.A => |*foo| {
            @compileLog(foo);
        },
    };
}
// 
```

The compiler also doesn't seem to catch the following syntax error:
```zig
comptime {
    const e: error{A}!u32 = error.A;

    if (e) |_| {} else |err| switch (err + "I can put anything I want here") {
        error.A => {},
    }
}
```

This is a little beyond me so I'm hoping that @Techatrix or another contributor knows what needs to be done here.

### Expected Behavior

should be a compile error like it was in 0.11

i forgot to demonstrate that this used to be an error in 0.11.0 but now triggers an assertion w/ a debug build of the compiler.  

```zig
// /tmp/tmp.zig
test {
    const e: error{A}!u32 = error.A;
    e catch |err| switch (err) {
        error.A => |*foo| {
            _ = foo;
        },
    };
}
```

```console
/tmp $ zig version
0.11.0
/tmp $ zig test /tmp/tmp.zig
/tmp/tmp.zig:38:21: error: error set cannot be captured by reference
        error.A => |*foo| {
                    ^~~~
/tmp $ zigup master-build # switch to local current debug compiler build
...
/tmp $ zig version
0.12.0-dev.2237+e09450214
/tmp $ zig test /tmp/tmp.zig
thread 498506 panic: reached unreachable code
~/dev/zig/zig/lib/std/debug.zig:403:14: 0x172eaac in assert (zig)
    if (!ok) unreachable; // assertion failure
             ^
~/dev/zig/zig/src/AstGen.zig:7221:19: 0x253aa16 in switchExprErrUnion (zig)
            const capture_token = case.payload_token orelse break :blk &err_scope.base;
                  ^
~/dev/zig/zig/src/AstGen.zig:1034:50: 0x1fc22d4 in expr (zig)
                        return switchExprErrUnion(gz, scope, ri.br(), node, .@"catch");
                                                 ^
~/dev/zig/zig/src/AstGen.zig:2518:41: 0x2a205fc in unusedResultExpr (zig)
    const maybe_unused_result = try expr(gz, scope, .{ .rl = .none }, statement);
                                        ^
~/dev/zig/zig/src/AstGen.zig:2499:65: 0x2a23197 in blockExprStmts (zig)
                else => noreturn_src_node = try unusedResultExpr(gz, scope, inner_node),
                                                                ^
~/dev/zig/zig/src/AstGen.zig:2310:27: 0x255f73a in blockExpr (zig)
        try blockExprStmts(&block_scope, &block_scope.base, statements);
                          ^
~/dev/zig/zig/src/AstGen.zig:1007:33: 0x1fc4a8a in expr (zig)
                return blockExpr(gz, scope, ri, node, statements[0..2]);
                                ^
~/dev/zig/zig/src/AstGen.zig:4762:34: 0x2523dc6 in testDecl (zig)
    const block_result = try expr(&fn_block, &fn_block.base, .{ .rl = .none }, body_node);
                                 ^
~/dev/zig/zig/src/AstGen.zig:5669:28: 0x1fb8787 in containerMember (zig)
            astgen.testDecl(gz, scope, wip_members, member_node) catch |err| switch (err) {
                           ^
~/dev/zig/zig/src/AstGen.zig:4981:49: 0x1c75e32 in structDeclInner (zig)
        var member = switch (try containerMember(&block_scope, &namespace.base, &wip_members, member_node)) {
                                                ^
~/dev/zig/zig/src/AstGen.zig:165:35: 0x19c5944 in generate (zig)
        if (AstGen.structDeclInner(
                                  ^
~/dev/zig/zig/src/Module.zig:2839:35: 0x20d395b in astGenFile (zig)
    file.zir = try AstGen.generate(gpa, file.tree);
                                  ^
~/dev/zig/zig/src/Compilation.zig:3824:19: 0x1dbeb6e in workerAstGenFile (zig)
    mod.astGenFile(file) catch |err| switch (err) {
                  ^
~/dev/zig/zig/lib/std/Thread/Pool.zig:94:39: 0x1af0b75 in runFn (zig)
            @call(.auto, func, closure.arguments);
                                      ^
~/dev/zig/zig/lib/std/Thread/Pool.zig:133:18: 0x202f527 in worker (zig)
            runFn(&run_node.data);
                 ^
~/dev/zig/zig/lib/std/Thread.zig:411:13: 0x1d2656a in callFn__anon_81706 (zig)
            @call(.auto, f, args);
            ^
~/dev/zig/zig/lib/std/Thread.zig:684:30: 0x1a55822 in entryFn (zig)
                return callFn(f, args_ptr.*);
                             ^
./nptl/pthread_create.c:442:8: 0x745f11e94ac2 in start_thread (pthread_create.c)
../sysdeps/unix/sysv/linux/x86_64/clone3.S:81:0: 0x745f11f2684f in ??? (../sysdeps/unix/sysv/linux/x86_64/clone3.S)
???:?:?: 0x0 in ??? (???)
Aborted (core dumped)
```

--- a/src/AstGen.zig
+++ b/src/AstGen.zig
@@ -841,13 +841,16 @@ fn expr(gz: *GenZir, scope: *Scope, ri: ResultInfo, node: Ast.Node.Index) InnerE
         .@"if",
         => {
             const if_full = tree.fullIf(node).?;
-            if (if_full.error_token) |error_token| {
-                const tag = node_tags[if_full.ast.else_expr];
-                if ((tag == .@"switch" or tag == .switch_comma) and
-                    std.mem.eql(u8, tree.tokenSlice(error_token), tree.tokenSlice(error_token + 4)))
-                {
-                    return switchExprErrUnion(gz, scope, ri.br(), node, .@"if");
+            no_switch_on_err: {
+                const error_token = if_full.error_token orelse break :no_switch_on_err;
+                switch (node_tags[if_full.ast.else_expr]) {
+                    .@"switch", .switch_comma => {},
+                    else => break :no_switch_on_err,
                 }
+                const switch_operand = node_datas[if_full.ast.else_expr].lhs;
+                if (node_tags[switch_operand] != .identifier) break :no_switch_on_err;
+                if (!mem.eql(u8, tree.tokenSlice(error_token), tree.tokenSlice(main_tokens[switch_operand]))) break :no_switch_on_err;
+                return switchExprErrUnion(gz, scope, ri.br(), node, .@"if");
             }
             return ifExpr(gz, scope, ri.br(), node, if_full);
         },
@@ -1026,16 +1029,21 @@ fn expr(gz: *GenZir, scope: *Scope, ri: ResultInfo, node: Ast.Node.Index) InnerE
         },
         .@"catch" => {
             const catch_token = main_tokens[node];
-            const payload_token: ?Ast.TokenIndex = if (token_tags[catch_token + 1] == .pipe) blk: {
-                if (token_tags.len > catch_token + 6 and
-                    token_tags[catch_token + 4] == .keyword_switch)
-                {
-                    if (std.mem.eql(u8, tree.tokenSlice(catch_token + 2), tree.tokenSlice(catch_token + 6))) {
-                        return switchExprErrUnion(gz, scope, ri.br(), node, .@"catch");
-                    }
+            const payload_token: ?Ast.TokenIndex = if (token_tags[catch_token + 1] == .pipe)
+                catch_token + 2
+            else
+                null;
+            no_switch_on_err: {
+                const capture_token = payload_token orelse break :no_switch_on_err;
+                switch (node_tags[node_datas[node].rhs]) {
+                    .@"switch", .switch_comma => {},
+                    else => break :no_switch_on_err,
                 }
-                break :blk catch_token + 2;
-            } else null;
+                const switch_operand = node_datas[node_datas[node].rhs].lhs;
+                if (node_tags[switch_operand] != .identifier) break :no_switch_on_err;
+                if (!mem.eql(u8, tree.tokenSlice(capture_token), tree.tokenSlice(main_tokens[switch_operand]))) break :no_switch_on_err;
+                return switchExprErrUnion(gz, scope, ri.br(), node, .@"catch");
+            }
             switch (ri.rl) {
                 .ref, .ref_coerced_ty => return orelseCatchExpr(
                     gz,

--- a/src/AstGen.zig
+++ b/src/AstGen.zig
@@ -7219,7 +7219,9 @@ fn switchExprErrUnion(
             };
 
             const capture_token = case.payload_token orelse break :blk &err_scope.base;
-            assert(token_tags[capture_token] == .identifier);
+            if (token_tags[capture_token] != .identifier) {
+                return astgen.failTok(capture_token + 1, "error set cannot be captured by reference", .{});
+            }
 
             const capture_slice = tree.tokenSlice(capture_token);
             if (mem.eql(u8, capture_slice, "_")) {

--- a/test/cases/compile_errors/error_set_cannot_capture_by_reference.zig
+++ b/test/cases/compile_errors/error_set_cannot_capture_by_reference.zig
@@ -1,0 +1,15 @@
+export fn entry() void {
+    const err: error{Foo} = error.Foo;
+
+    switch (err) {
+        error.Foo => |*foo| {
+            foo catch {};
+        },
+    }
+}
+
+// error
+// backend=stage2
+// target=native
+//
+// :5:23: error: error set cannot be captured by reference

--- a/test/cases/compile_errors/switch_on_error_union_discard.zig
+++ b/test/cases/compile_errors/switch_on_error_union_discard.zig
@@ -1,9 +1,7 @@
 export fn entry() void {
     const x: error{}!u32 = 0;
-    if (x) |v| v else |_| switch (_) {
-    }
+    if (x) |v| v else |_| switch (_) {}
 }
-
 
 // error
 // backend=stage2

--- a/test/cases/compile_errors/switch_on_error_with_capture_by_reference.zig
+++ b/test/cases/compile_errors/switch_on_error_with_capture_by_reference.zig
@@ -1,0 +1,24 @@
+comptime {
+    const e: error{Foo}!u32 = error.Foo;
+    e catch |err| switch (err) {
+        error.Foo => |*foo| {
+            foo catch {};
+        },
+    };
+}
+
+comptime {
+    const e: error{Foo}!u32 = error.Foo;
+    if (e) {} else |err| switch (err) {
+        error.Foo => |*foo| {
+            foo catch {};
+        },
+    }
+}
+
+// error
+// backend=stage2
+// target=native
+//
+// :4:24: error: error set cannot be captured by reference
+// :13:24: error: error set cannot be captured by reference

--- a/test/cases/compile_errors/switch_on_error_with_non_trivial_switch_operand.zig
+++ b/test/cases/compile_errors/switch_on_error_with_non_trivial_switch_operand.zig
@@ -1,0 +1,22 @@
+export fn entry1() void {
+    var x: error{Foo}!u32 = 0;
+    _ = &x;
+    if (x) |_| {} else |err| switch (err + 1) {
+        else => {},
+    }
+}
+
+export fn entry2() void {
+    var x: error{Foo}!u32 = 0;
+    _ = &x;
+    _ = x catch |err| switch (err + 1) {
+        else => {},
+    };
+}
+
+// error
+// backend=stage2
+// target=native
+//
+// :4:42: error: invalid operands to binary expression: 'ErrorSet' and 'ComptimeInt'
+// :12:35: error: invalid operands to binary expression: 'ErrorSet' and 'ComptimeInt'

